src/middlewared/middlewared/main.py

```diff
@@ -183,11 +183,17 @@ def send_error(self, message, errno, reason=None, exc_info=None, etype=None, ext
 
             if self.middleware.crash_reporting.is_disabled():
                 self.logger.debug('[Crash Reporting] is disabled using sentinel file.')
+            elif self.middleware.crash_reporting_semaphore.locked():
+                self.logger.debug('[Crash Reporting] skipped due too many running instances')
             else:
-                extra_log_files = (('/var/log/middlewared.log', 'middlewared_log'),)
-                asyncio.ensure_future(self.middleware.run_in_thread(
-                    self.middleware.crash_reporting.report, sys.exc_info(), None, extra_log_files
-                ))
+                async with self.middleware.crash_reporting_semaphore:
+                    extra_log_files = (('/var/log/middlewared.log', 'middlewared_log'),)
+                    await self.middleware.run_in_io_thread(
```

Comment:
IMO `await` is not appropriate here. I think we can put all this new code (checking if locked, acquire, sending crash report) into a coroutine and just ensure_future here